runtime/doc/lsp.txt

```diff
@@ -0,0 +1,359 @@
+
+Neovim only exposes a powerful API for the language server protocol. To get
+the real benefits of this API, a language server must be installed. Many
+examples can be found here:
+
+	https://microsoft.github.io/language-server-protocol/implementors/servers/
+
+After installing a language server to your machine, you must let Neovim know
+how to start and interact with that language server. To do so, use the
+following APIs:
+
+                                                      *lsp#add_server_config()*
+lsp#add_server_config({config}):
+	Used to add a language server.
+
+	This will add an autocmd that will start the server and handle any other
+	startup activities whenever the |FileType| event is triggered for {ftype}.
+
+	Parameters: ~
+		{config} (dictionary): 
+			`filetype` (string|array):
+				A string or list of strings of filetypes to
+				associate with this server.
+
+			`server_name` (string) (optional):
+				You can pass any name as server_name.
+				If you want to use multiple language servers
+				for one filetype, be sure to add server_name.
+
+			`cmd` (dictionary): Executing language server command.
+				`execute_path` (string):
+					Executing language server command path.
+				`args` (array):
+					Command arguments.
+
+			`server_config` (dictionary): Optional configuration for a server.
+
+	Example: ~
+>
+		" C, C++ example
+		call lsp#add_server_config({
+		      \ 'filetype': ['c', 'cpp'],
+		      \ 'server_name': 'cland',
+		      \ 'cmd': { 'execute_path': 'cland', 'args': ['-background-index'] }})
+<
+>
+		" Go example
+		call lsp#add_server_config({
+		      \ 'filetype': 'go',
+		      \ 'server_name': 'gopls',
+		      \ 'cmd': { 'execute_path': 'gopls', 'args': [] }})
+<
+>
+		" Python example
+		call lsp#add_server_config({
+		      \ 'filetype': 'python',
+		      \ 'server_name': 'pyls',
+		      \ 'cmd': { 'execute_path': 'pyls', 'args': [] }})
+<
+>
+		" Rust example
+		call lsp#add_server_config({
+		      \ 'filetype': 'rust',
+		      \ 'server_name': 'rls',
+		      \ 'cmd': { 'execute_path': 'rustup', 'args': ['run', 'stable', 'rls'] },
+		      \ 'server_config': {
+		      \   'clippy_preference': 'on',
+		      \   'all_targets': v:false,
+		      \   'build_on_save': v:true,
+		      \   'wait_to_build': 0
+		      \ }})
+<
+
+================================================================================
+                                                                  *lsp-request*
+Requests
+
+To make a request to the language server client running in Nvim, the following
+API is defined:
+>
+	:lua vim.lsp.request(method, arguments, callback, bufnr, filetype)
+	:lua vim.lsp.request_async(method, arguments, callback, bufnr, filetype)
+<
+
+                                                            *vim.lsp.request()*
+vim.lsp.request({method} [, {arguments} [, {callback} [, {bufnr} [, {filetype} [, {server_name}]]]]]):
+	Creates and sends a request to the server for [filetype]. Will fill in any
```

Comment:
[x] to {x} for args in the signature. Here, for `filetype` and below for `callback`.